src/coins.cpp

```diff
@@ -330,6 +330,9 @@ bool CCoinsViewCache::HaveInputs(const CTransaction& tx) const
 {
     if (!tx.IsCoinBase()) {
         for (unsigned int i = 0; i < tx.vin.size(); i++) {
+            if (tx.vin[i].m_is_pegin) {
```

Comment:
there's no risk with this? essentially to say that we "have" every peg-in input, regardless of if the peg-in is valid or not. (I assume "having" an input normally implies that it's valid/available)

Reply:
"Have" usually means it exists in the UTXO database, but for peg-in inputs we aren't going to look there. Instead we are going to do two things:
1) Look for a well-formed pegin witness
2) Make sure that peg-in has not already occurred by looking for it in our "spent peg-in" database.